host/webhooks/webhooks.go

```diff
@@ -0,0 +1,252 @@
+
+type (
+	WebhookStore interface {
+		DeleteWebhook(wh Webhook) error
+		AddWebhook(wh Webhook) error
+		Webhooks() ([]Webhook, error)
+	}
+
+	Broadcaster interface {
+		BroadcastAction(ctx context.Context, action Event) error
+	}
+)
+
+const (
+	webhookTimeout   = 10 * time.Second
+	WebhookEventPing = "ping"
+)
+
+type Manager struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+	wg        sync.WaitGroup
+	store     WebhookStore
+
+	mu       sync.Mutex
+	queues   map[string]*eventQueue // URL -> queue
+	webhooks map[string]Webhook
+}
+
+type eventQueue struct {
+	ctx context.Context
+	url string
+
+	mu           sync.Mutex
+	isDequeueing bool
+	events       []Event
+}
+
+func (w *Manager) Close() error {
+	w.ctxCancel()
+	w.wg.Wait()
+	return nil
+}
+
+func (w Webhook) String() string {
+	return fmt.Sprintf("%v.%v.%v", w.URL, w.Module, w.Event)
+}
+
+func (w *Manager) Delete(wh Webhook) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	// if err := w.store.DeleteWebhook(wh); errors.Is(err, gorm.ErrRecordNotFound) {
+	// 	return ErrWebhookNotFound
+	// } else if err != nil {
+	// 	return err
+	// }
+	err := w.store.DeleteWebhook(wh)
+	if err != nil {
+		return err
+	}
+	delete(w.webhooks, wh.String())
+	return nil
+}
+
+func (w *Manager) Info() ([]Webhook, []WebhookQueueInfo) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	var hooks []Webhook
+	for _, hook := range w.webhooks {
+		hooks = append(hooks, Webhook{
+			Event:  hook.Event,
+			Module: hook.Module,
+			URL:    hook.URL,
+		})
+	}
+	var queueInfos []WebhookQueueInfo
+	for _, queue := range w.queues {
+		queue.mu.Lock()
+		queueInfos = append(queueInfos, WebhookQueueInfo{
+			URL:  queue.url,
+			Size: len(queue.events),
+		})
+		queue.mu.Unlock()
+	}
+	return hooks, queueInfos
+}
+
+func (a Event) String() string {
+	return a.Module + "." + a.Event
+}
+
+func (w *Manager) BroadcastAction(_ context.Context, event Event) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	for _, hook := range w.webhooks {
+		if !hook.Matches(event) {
+			continue
+		}
+
+		// Find queue or create one.
+		queue, exists := w.queues[hook.URL]
+		if !exists {
+			queue = &eventQueue{
+				ctx: w.ctx,
+				url: hook.URL,
+			}
+			w.queues[hook.URL] = queue
+		}
+
+		// Add event and launch goroutine to start dequeueing if necessary.
+		queue.mu.Lock()
+		queue.events = append(queue.events, event)
+		if !queue.isDequeueing {
+			queue.isDequeueing = true
+			w.wg.Add(1)
+			go func() {
+				queue.dequeue()
+				w.wg.Done()
+			}()
+		}
+		queue.mu.Unlock()
+	}
+	return nil
+}
+
+func (q *eventQueue) dequeue() {
+	for {
+		q.mu.Lock()
+		if len(q.events) == 0 {
+			q.isDequeueing = false
+			q.mu.Unlock()
+			return
+		}
+		next := q.events[0]
+		q.events = q.events[1:]
+		q.mu.Unlock()
+
+		err := sendEvent(q.ctx, q.url, next)
+		if err != nil {
+			//fmt.Errorf("failed to send Webhook event %v to %v: %v", next.String(), q.url, err)
+		}
+	}
+}
+
+func (w Webhook) Matches(action Event) bool {
```

Comment:
I'm not sure how I'd check this without something similar to the Matches function here. I've updated this function in commit [deb27da](https://github.com/SiaFoundation/hostd/pull/192/commits/deb27daebff424b380bd5d59d377001798fda7d3) to use scope, but I couldn't figure out a way to avoid having this function.